resources/type-system/upper-lower-bounds.md

```diff
@@ -162,22 +166,38 @@ We define the upper bound of two types T1 and T2 to be **UP**(`T1`,`T2`) as foll
 - **UP**(`X1 extends B1`, `T2`) =
   - `T2` if `X1 <: T2`
   - otherwise `X1` if `T2 <: X1`
-  - otherwise **UP**(`B1[Object/X1]`, `T2`)
+  - otherwise **UP**(`B1a`, `T2`)
+    where `B1a` is obtained from `B1` by replacing contravariant occurrences
```

Comment:
I'm not quite parsing this text well.  I think what we want here is something like:

```
otherwise **UP(`B1a`, `T2`) where `B1a` is the greatest closure of `B1` with respect to `X1` as defined in inference.md
```

WDYT?

Reply:
Yes, I think that will work, based on the assumption that we first expand all type aliases (so invariant occurrences of a type variable is simply an occurrence in the bound of a type parameter of a function type).

With declaration-site variance added in we can't expand away all the invariant type parameters, but we would then revisit the definition of 'greatest closure of a type'.

Reply:
Done.